documentation/docs/explanations/concurrency.md

```diff
@@ -0,0 +1,80 @@
+
+This approach advantage is that the idea is relatively simple and implement the STM model. The downside is that the implementation we developed of TL2 is painfully slow and usually going single thread is faster than using multiple threads. 
+We suspect that this poor performance are due to multiple interleaving of threads which force the transactions to be reset multiple times. Further investigation is required to confirm this idea.
+
+
+## Read-Log-Update (RLU)
+
+RLU is an extension of the more famous _Read-Copy-Update_ (RCU) that has been widely adopted in the Linux kernel.
+
+RLU was first presented in 2015 in the paper [Read-Log-Update: A Lightweight Synchronization Mechanism for Concurrent Programming](http://sigops.org/sosp/sosp15/current/2015-Monterey/printable/077-matveev.pdf). 
+
+Contrary to TL2 previously presented, RLU is a blocking algorithm.
+Threads do their computations locally and synchronize with each other using their local clock (when the computations started) and a common global clock.
+A thread logs the new value of any shared memory that it has modified.
+To read shared memory a thread either does it directly in the memory or go fetch it in another thread logs when it has been locked and is being modified.
+When a thread wants to commit its computations to the memory it has to wait that all the other threads that have started with an old memory state to terminate beforehand.
+The upside of such approach is that a thread never needs to redo its computation since it has waited the correct timing to commit its work.
+In return threads have to spend some time idle since they have to wait. 
+However the whole RLU algorithm is lock-free meaning that at any moment there is always a thread that advances in its computation and the whole system cannot be hard stuck.
+
+## Concurrent programming with locks
+
+Using locks is one of the most basic approach to concurrent programming. 
+Lock are used to control the access to shared memory/variables. Depending of the type of locks used you can restrict the access to a shared memory/variable to either one or a limited number of threads. 
+This prevents data race since the different threads cannot access the shared memory/variables in a chaotic way. 
+The biggest issue is that those locks are in general difficult to use correctly and really hard to debug. 
+One of the most problematic situation that can happen with locks are __deadlocks__. 
+Deadlocks happen when the whole system cannot advance anymore due to the fact that different threads require some locks to advance their computation but these locks are kept and blocked by other threads in a similar situation.
+Models like the actor system or the STM have been created to use locks as little as possible.
+
+In our case the API of Stronghold is relatively straightforward therefore we're exploring the idea of using basic locks instead of a resource heavy concurrent framework.
```

Comment:
```suggestion
Using locks is one of the most basic approaches to concurrent programming.
Locks are used for controlling access to shared memory/variables. Depending on the type of locks used, you can restrict access to a shared memory/variable to either one or a limited number of threads.
This prevents data race since the different threads cannot access the shared memory/variables chaotically.
The biggest issue is that those locks are generally difficult to use correctly and hard to debug.

One of the most problematic situations with locks are __deadlocks__.
Deadlocks happen when the whole system cannot advance anymore because different threads require some locks to advance their computation, but these locks are kept and blocked by other threads in a similar situation.
Models like the [actor system](the-actor-model) or the [STM](#software-transactional-memory-stm-with-transactional-locking-2-tl2) use locks as little as possible.

In our case, Stronghold’s API is relatively straightforward, so we're exploring the idea of using basic locks instead of a resource-heavy concurrent framework.
```